eventrecorder/recorder.go

```diff
@@ -147,99 +151,128 @@ func (r *EventRecorder) handleRetrievalEvents(res http.ResponseWriter, req *http
+}
+
+func handleQueryAskEvent() {
+	stats.Record(context.Background(), metrics.RequestWithSuccessfulQueriesCount.M(1))
+}
+
+// handleFailureEvent is called when a query _or_ retrieval fails
+func handleFailureEvent(event Event) {
+
+	detailsObj, ok := event.EventDetails.(map[string]interface{})
+	if !ok {
 		return
 	}
-	err = json.NewEncoder(res).Encode(summary)
-	if err != nil {
-		http.Error(res, err.Error(), http.StatusInternalServerError)
-		logger.Errorw("failed encoding result", "err", err)
+	msg, ok := detailsObj["Error"].(string)
+	if !ok {
 		return
 	}
+	switch event.Phase {
+	case types.QueryPhase:
+		var matched bool
+		for substr, metric := range metrics.QueryErrorMetricMatches {
+			if strings.Contains(msg, substr) {
+				stats.Record(context.Background(), metric.M(1))
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			stats.Record(context.Background(), metrics.QueryErrorOtherCount.M(1))
+		}
+	case types.RetrievalPhase:
+		stats.Record(context.Background(), metrics.RetrievalDealFailCount.M(1))
+		stats.Record(context.Background(), metrics.RetrievalDealActiveCount.M(-1))
 
-	r.db.Exec(ctx, "TRUNCATE TABLE retrieval_events")
-	logger.Infow("Successfully ran summary and cleared DB")
+		var matched bool
+		for substr, metric := range metrics.ErrorMetricMatches {
+			if strings.Contains(msg, substr) {
+				stats.Record(context.Background(), metric.M(1))
+				matched = true
+				break
+			}
+		}
+		if !matched {
+			stats.Record(context.Background(), metrics.RetrievalErrorOtherCount.M(1))
+		}
+	}
 }
 
-func (r *EventRecorder) handleReady(res http.ResponseWriter, req *http.Request) {
-	switch req.Method {
-	case http.MethodGet:
-		// TODO: ping DB as part of readiness check?
-		res.Header().Add("Allow", http.MethodGet)
-	default:
-		http.Error(res, "", http.StatusMethodNotAllowed)
+func handleStartedEvent(event Event) {
+	if event.Phase == types.RetrievalPhase {
+		stats.Record(context.Background(), metrics.RetrievalRequestCount.M(1))
+		stats.Record(context.Background(), metrics.RetrievalDealActiveCount.M(1))
 	}
 }
 
-func (r *EventRecorder) Shutdown(ctx context.Context) error {
-	return r.server.Shutdown(ctx)
+func handleCandidatesFilteredEvent(event Event) {
+	detailsObj, ok := event.EventDetails.(map[string]interface{})
+	if !ok {
+		return
+	}
+
+	candidateCount, ok := detailsObj["CandidateCount"].(float64)
+	if !ok {
+		return
+	}
+	if candidateCount > 0 {
+		stats.Record(context.Background(), metrics.RequestWithIndexerCandidatesFilteredCount.M(1))
+	}
+}
+
+func handleCandidatesFoundEvent(event Event) {
+	detailsObj, ok := event.EventDetails.(map[string]interface{})
+	if !ok {
+		return
+	}
+
+	candidateCount, ok := detailsObj["CandidateCount"].(float64)
+	if !ok {
+		return
+	}
+	if candidateCount > 0 {
+		stats.Record(context.Background(), metrics.RequestWithIndexerCandidatesCount.M(1))
+	}
+	stats.Record(context.Background(), metrics.IndexerCandidatesPerRequestCount.M(int64(candidateCount)))
```

Comment:
What is the difference between `IndexerCandidatesPerRequestCount` and `RequestWithIndexerCandidatesCount`?

Reply:
For `RequestWithIndexerCandidatesCount` we are measuring for the increasingly shrinking funnel of retrieval requests coming in at the top and successful retrievals happening at the bottom, this counter ticks over when we get candidates and can move on to the next part of the funnel.

`IndexerCandidatesPerRequestCount` counts the number of candidates received for the request so we can do things like average number of candidates per request. Probably not as valuable as the former now but makes for interesting graphs.